schematic/schemas/create_json_schema.py

```diff
@@ -0,0 +1,895 @@
+    def _update_reverse_dependencies(
+        self, node_display_name: str, node_dependencies_display_names: list[str]
+    ) -> None:
+        """Updates the reverse dependencies
+
+        Arguments:
+            node_display_name: The display name of the node
+            node_dependencies_display_names: the display names of the reverse dependencies
+        """
+        for dep in node_dependencies_display_names:
+            if dep not in self._reverse_dependencies:
+                self._reverse_dependencies[dep] = []
+            self._reverse_dependencies[dep].append(node_display_name)
+
+    def _update_nodes_to_process(self, nodes: list[str]) -> None:
+        """Updates the nodes to process with the input nodes
+
+        Arguments:
+            nodes: Nodes to add
+        """
+        self._nodes_to_process += nodes
+
+
+def create_json_schema(  # pylint: disable=too-many-arguments
+    dmge: DataModelGraphExplorer,
+    datatype: str,
+    schema_name: str,
+    write_schema: bool = True,
+    schema_path: Union[str, None] = None,
+    jsonld_path: Optional[str] = None,
+    use_property_display_names: bool = True,
+    use_valid_value_display_names: bool = True,
+) -> dict[str, Any]:
+    """
+    Creates a JSONSchema dict for the datatype in the data model.
+
+    This uses the input graph starting at the node that corresponds to the input datatype.
+    Starting at the given node it will(recursively):
+    1. Find all the nodes this node depends on
+    2. Find all the allowable metadata values / nodes that can be assigned to a particular
+        node (if such a constraint is specified on the schema).
+
+    Using the above data it will:
+    - Cerate properties for each attribute of the datatype.
+    - Create properties for attributes that are conditionally
+        dependent on the datatypes attributes
+    - Create conditional dependencies linking attributes to their dependencies
+
+    Arguments:
+        dmge: A DataModelGraphExplorer with the data model loaded
+        datatype: the datatype to create the schema for.
+            Its node is where we can start recursive dependency traversal
+            (as mentioned above).
+        schema_name: Name assigned to JSON-LD schema (to uniquely identify it via URI
+            when it is hosted on the Internet).
+        write_schema: whether or not to write the schema as a json file
+        schema_path: Where to save the JSON Schema file
+        jsonld_path: Used to name the file if the path isn't supplied
+        use_property_display_names: If True, the properties in the JSONSchema
+          will be written using node display names
+        use_valid_value_display_names: If True, the valid_values in the JSONSchema
+          will be written using node display names
+
+    Returns:
+        JSON Schema as a dictionary.
+    """
+    graph_state = GraphTraversalState(dmge, datatype)
+
+    json_schema = JSONSchema(
+        schema_id="http://example.com/" + schema_name,
```

Comment:
Will this be filled in from somewhere eventually instead of using the static `http://example.com/`?

Reply:
AFAIK this field has no effect on JSON Schema validaiton. So it's really up to what users want. @thomasyu888 What do you think? This becomes the `$id` in the final schema. It's described [here](https://json-schema.org/learn/getting-started-step-by-step#create-a-schema-definition) as :+1: 

```
$id: sets a URI for the schema. You can use this unique URI to refer to elements of the schema from inside the same document or from external JSON documents.
```

Should we allow users to set it via the CLI?

Reply:
Let's not allow it until enough people ask for it. Thanks!